client/config_test.go

```diff
@@ -0,0 +1,274 @@
+			expStatus:  http.StatusOK,
+		},
+		{
+			name:       "TestConfigDeleteNotFound",
+			function:   "delete",
+			key:        model.Key{Name: "name", Namespace: "namespace", Kind: "route-rule"},
+			expError:   errNotFound,
+			expHeaders: http.Header{"Content-Type": []string{"text/plain"}},
+			expStatus:  http.StatusNotFound,
+		},
+		{
+			name:       "TestConfigDeleteInvalidConfigType",
+			function:   "delete",
+			key:        model.Key{Name: "name", Namespace: "namespace", Kind: "route-rule"},
+			expError:   errInvalidType,
+			expHeaders: http.Header{"Content-Type": []string{"text/plain"}},
+			expStatus:  http.StatusBadRequest,
+		},
+		{
+			name:      "TestConfigListWithNamespace",
+			function:  "list",
+			kind:      "kind",
+			namespace: "namespace",
+			expConfigSlice: []apiserver.Config{
+				apiserver.Config{Type: "type", Name: "name", Spec: "spec"},
+				apiserver.Config{Type: "type", Name: "name2", Spec: "spec"},
+			},
+			expHeaders: http.Header{"Content-Type": []string{"application/json"}},
+			expStatus:  http.StatusOK,
+		},
+		{
+			name:     "TestConfigListWithoutNamespace",
+			function: "list",
+			kind:     "kind",
+			expConfigSlice: []apiserver.Config{
+				apiserver.Config{Type: "type", Name: "name", Spec: "spec"},
+				apiserver.Config{Type: "type", Name: "name2", Spec: "spec"},
+			},
+			expHeaders: http.Header{"Content-Type": []string{"application/json"}},
+			expStatus:  http.StatusOK,
+		},
+		{
+			name:       "TestConfigListWithNamespaceInvalidConfigType",
+			function:   "list",
+			kind:       "kind",
+			namespace:  "namespace",
+			expError:   errInvalidType,
+			expHeaders: http.Header{"Content-Type": []string{"application/json"}},
+			expStatus:  http.StatusBadRequest,
+		},
+	}
+	for _, c := range cases {
+
+		// Setup test server
+		var exp string
+		if c.expError != nil {
+			exp = c.expError.Error()
+		} else {
+			if c.expConfigSlice != nil {
+				e, _ := json.Marshal(c.expConfigSlice)
+				exp = string(e)
+			} else {
+				e, _ := json.Marshal(c.expConfig)
+				exp = string(e)
+			}
+
+		}
+		fh := &FakeHandler{
+			expResponse:   exp,
+			expHeaders:    c.expHeaders,
+			expStatusCode: c.expStatus,
+		}
+		ts := httptest.NewServer(http.HandlerFunc(fh.HandlerFunc))
+		defer ts.Close()
+		tsURL, _ := url.Parse(ts.URL)
+
+		// Setup Client
+		var config *apiserver.Config
+		var configSlice []apiserver.Config
+		var err error
+		client := NewManagerClient(tsURL, "test", nil)
+		switch c.function {
+		case "get":
+			config, err = client.GetConfig(c.key)
+		case "add":
+			err = client.AddConfig(c.key, *c.config)
+		case "update":
+			err = client.UpdateConfig(c.key, *c.config)
+		case "delete":
+			err = client.DeleteConfig(c.key)
+		case "list":
+			configSlice, err = client.ListConfig(c.kind, c.namespace)
+		default:
+			t.Fatal("didn't supply function to test case, don't know which client function to call")
+		}
+
+		// Verify
+		if c.expError == nil && err != nil {
```

Comment:
It would be useful to include the test name in each of these failures so developers don't need to inspect the source (or re-run the test) to understand the specific test failure.